recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,401 @@
++        ]
++        envSecretsFrom: {               // All Secrets from Secret Store
++           source: secretStoreOCICred.id
++         }        
++      }      
++    }
+  }
+  recipes: {      
+    ...
+  }
+}
+
+Option 2:
+This option provides grouping and efficiency to retrieve all details for a single provider.
+
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'dsrp-resources-env-recipes-context-env'
+  location: 'global'
+  properties: {
+    compute: {
+      ...
+    }
+    providers: {
+      ...
+    }
+    recipeConfig: {
+      terraform: {
+        authentication:{
+          ...        
+        }
++       providers: {
++         'azurerm': [
++           {
++             subscriptionid: 1234,
++             secrets: {
++               source: 'secretStoreAz.id'
++            }
++          },
++          {
++             subscriptionid: 1234,
++             alias: 'az-paymentservice',
++             secrets: {
++               source: 'secretStoreAzPayment.id'
++            }
++          }]
++          'gcp': [
++            {
++              project: 1234,
++              regions: ['us-east1', 'us-west1']
++            }
++          ]
++          'oraclepass': [
++            {
++              database_endpoint: "...",
++              java_endpoint: "...",
++              mysql_endpoint: "..."
++            }
++          ]
++        }
++     }
++     env: {
++       key: value
++       envSecrets: [                     // Individual Secrets from SecretStore
++         {
++            name: 'SPECIAL_LEVEL_KEY'
++            valueFrom: {
++              source: secretStoreConfig.id
++              name: 'special-config'
++              key: 'special.level'
++           }
++         },
++         {
++            name: 'LOG_LEVEL'
++            valueFrom: {
++              source: secretStoreConfig.id
++              name: 'special-config'
++              key: 'special.log'
++           }
++         }
++        ]
++        envSecretsFrom: {                  // All Secrets from Secret Store
++           source: secretStoreOCICred.id
++        }
++     }
+    }
+  }
+  recipes: {      
+    ...
+  }
+}
+
+
+```
+
+
+Limitations: Customers may store sensitive data in other formats which may not be supported. eg. sensitive data is stored on files, which customers will not currently be able to load on disk in applications-rp where tf init/apply commands are run.
+Containerization work may alleviate this limitations. Further design work will be needed for towards this which is planned for the near future. 
+
+### API design (if applicable) (TBD)
+
+***Model changes providers***
+
+### Option 1:
+```
+Addition of new property to TerraformConfigProperties in `recipeConfig` under environment properties.
+
+model TerraformConfigProperties{
+  @doc(Specifies authentication information needed to use private terraform module repositories.)  
+  authentication?: AuthConfig
++ providers?: Array<ProviderConfig>
+}
```

Comment:
Should we make this more generic so that we can use this for Bicep later too?

Reply:
TerraformConfigProperties currently exists under recipeConfig, we should be able to extend that later into Bicep specific provider configuration as needed.